Add a userscript name formatter arg to build()

atj opened the pull request:
This allows customising the name of the userscript added to each heading in `README.md`. The default value calls `camelToTitleCase(baseName)` to preserve backward compatibility.

atj committed:
Add userscriptNameFormatter argument to build()

This enables the caller to customise how the userscript name is
formatted in the generated `README.md` file. The default name formatter
returns the value of `metadata.name`.

atj commented:
I'm terrible at naming variables so if you have any better suggestions please let me know.

I have not tested this yet but wanted to see what you thought about the approach before I did so.

kellnerd commented:
Thanks for the PR!

> I'm terrible at naming variables so if you have any better suggestions please let me know.

I actually prefer to have clear parameter names, even if they are verbose. So I think that the name `userscriptNameFormatter` is a good choice.

> The default value calls camelToTitleCase(baseName) to preserve backward compatibility.

At this stage I also don't mind breaking changes, there's no need to have a weird, non-intuitive default just because my setup is currently using it.

Also the current function interfaces were not really a well thought-out design decision, I've simply converted things which had previously been constants into parameters. (Yesterday I noticed that there's not even a parameter to configure the target directory, it is always `dist`.)
To be honest, the only good decision was to use object destructuring for `build()` instead of an ordered parameter list (which is still used by e.g. `buildUserscript()`), that gives us more flexibility to add parameters or make them optional.

It would be a good idea to also use this approach for the name formatter:

```js
// default parameter value:
userscriptNameFormatter = ({ baseName, metadata }) => metadata.name
// to mimic the current behaviour:
userscriptNameFormatter = ({ baseName }) => camelToTitleCase(baseName)
// for my repo, now that you've reminded how unintuitive the current approach is:
userscriptNameFormatter = ({ metadata }) => metadata.name.replace(/^MusicBrainz: /, '')

// these can all be called with:
userscriptNameFormatter({ baseName, metadata })
```

Probably the default function should not be defined inline, but as a regular function, which allows us to add JSDoc type annotations.

---

One additional thing which also has to be adapted is the homepage URL generator
https://github.com/kellnerd/userscript-bundler/blob/ef3a35dc28425f601ee1e76f3e2e08c07757da78/src/userscriptMetadata.js#L43
... which also relies on the heading being derived from the basename:
https://github.com/kellnerd/userscript-bundler/blob/ef3a35dc28425f601ee1e76f3e2e08c07757da78/src/github.js#L37-L43

Since this requires passing the formatter function down to `buildUserscript`, it finally forces a refactoring of the function's interface... Just let me know whether you want to deal with that or rather not, then I will do that myself.

If we also use an object parameter here, we could even think about grouping the parameters of `build` logically so that they can be directly passed to the downstream functions:

```js
export async function build({
	userscripts = {
		sourcePath: 'src/userscripts',
		targetPath: 'dist',
		nameFormatter: defaultNameFormatter,
	},
	bookmarklets = false, // or: { sourcePath: 'src/bookmarklets', ... }
	readmePath = 'README.md',
	docBasePath = 'doc',
	debug = false,
} = {}) {
	// build userscripts
	const userscriptNames = await buildUserscripts({ ...userscripts, debug });
	...
}
```

But I haven't tested that yet, maybe it's too complex and the default values also don't work as smooth as previously.

P.S. Sorry for the long comment, but this feature remembered me of quite a few things which still can be improved.

atj commented:
Absolutely no need to apologise for leaving a detailed comment - I appreciate you taking the time!

My understanding of modern JS isn't great and whilst I've already learned a lot from reading/using your userscripts and building my own using your exports, this sort of feedback is very helpful as it gives me some useful context about how best to design / structure reusable JS.

I have to say that I find the lack of keyword arguments for functions quite irritating, so the use of object parameters appeals to me. I'll update the PR, but I think we'll probably need a few iterations so bare with me!

atj commented:
OK, I've updated the PR. To be honest I feel a bit out of my depth - I haven't tested it yet so I'm not even sure this code works, but VSCode isn't complaining and seems to be picking up the type signatures correctly.

I won't be offended if this is _so_ far from being correct you think it'd be easier to just implement it yourself, but I'd like to try and get to a merge-able PR if possible!

kellnerd commented on src/build.js:
```diff
@@ -9,15 +9,27 @@ import { getMarkdownFiles } from './getFiles.js'
-	userscriptBasePath = 'src/userscripts',
+	userscripts = {
+		sourcePath: 'src/userscripts',
+		destPath: 'dist',
```
Allowing to change the destination path (which is a better wording than target path, thanks!) requires some additional changes in `buildUserscript` (singular) and the class `GitRepo` would have to be instantiated dynamically with the correct value instead of `'dist'`.
Do you also want to do these changes or should we drop the parameter for now?

atj replied:
I've dropped it for now.

kellnerd commented on src/build.js:
```diff
@@ -9,15 +9,27 @@ import { getMarkdownFiles } from './getFiles.js'
+	userscripts = {
+		sourcePath: 'src/userscripts',
+		destPath: 'dist',
+		nameFormatter: defaultNameFormatter,
```
There is one problem with this nested parameter: The fallback values only work for the first level of the parameter object which will be destructured and has defaults specified with `=`.
Unfortunately this syntax can't be used for the inner object, so the following won't work:
```js
export async function build({
	bookmarkletBasePath = false,
	userscriptBasePath = 'src/userscripts',
	userscripts = {
		sourcePath = 'src/userscripts',
		destPath = 'dist',
		nameFormatter = defaultNameFormatter,
	},
	...
} = {})
```
So we either have to provide defaults in the function body or discard the approach to use nested parameters.

atj replied:
How about we discard the nested parameter approach and go back to prefixing the variable names with `userscript` instead? I think defining defaults inside functions is opaque and reduces maintainability. Also the main benefit of object destructuring for me is that we can define defaults in the function header and the caller only needs to specify the arguments they want to change.

So something like:

```js
export async function build({
	userscriptSourcePath = 'src/userscripts',
	userscriptDestPath = 'dist',
	userscriptNameFormatter = defaultNameFormatter,
	bookmarkletSourcePath = 'src/bookmarklets',
	docSourcePath = 'doc',
	readmePath = 'README.md',
	debug = false,
} = {})
```

kellnerd replied:
Yeah, that was also my thought after testing the behaviour of nested parameters, good to have it confirmed. It's better to give up the idea of passing an entire parameter object down to an underlying function, having to reassign multiple parameters to `buildUserscripts` is still better than having to deal with fallback values in the function body again.

atj committed:
build(): s/BasePath/SourcePath/g

atj committed:
Add documentation and typedefs for build function

This commit also changes the default value for bookmarkletSourcePath
from `false` to `null`, which simplifies the type annotation and is
arguably more correct.

atj commented:
OK, hopefully this is a bit closer to the final product...

I'm not sure if the TypeScript definition for `UserscriptNameFormatter` is needed? Is it possible to define a type in JSDoc comments?

I also changed the default value for `bookmarkletSourcePath` to `null` from `false`, as this seems a more correct way of specifying "no value".

I've performed some very basic testing generating a single userscript but no bookmarklets.

kellnerd commented on src/types/UserscriptNameFormatter.d.ts:
```diff
@@ -0,0 +1,3 @@
+import { UserscriptMetadata } from "./UserscriptMetadata"
```
```suggestion
import type { UserscriptMetadata } from './UserscriptMetadata';
```
Minor nitpick for consistency...
I know that VS Code defaults to double quotes unless the file already contains imports with single quotes (or you've changed the default setting).

P.S. Might be worth renaming this file to `BuildOptions.d.ts`, I could imagine having additional types for the other build functions (userscripts and bookmarklets) soon / in the future.

atj replied:
Fixed and renamed to `BuildOptions.d.ts`.

kellnerd commented on src/build.js:
```diff
@@ -9,40 +9,59 @@ import { getMarkdownFiles } from './getFiles.js'
+ * Returns the formatted name of the userscript for use in documentation.
+ * @type {import('./types/UserscriptNameFormatter.js').UserscriptNameFormatter}
+ */
+function defaultNameFormatter({metadata}) {
```
```suggestion
function defaultNameFormatter({ metadata }) {
```

atj replied:
Fixed.

kellnerd commented on src/build.js:
```diff
@@ -9,40 +9,59 @@ import { getMarkdownFiles } from './getFiles.js'
 		const metadata = await loadMetadata(filePath);
 
-		readme.write(`\n### ${camelToTitleCase(baseName)}\n`);
+		readme.write(`\n### ${userscriptNameFormatter({baseName, metadata})}\n`);
```
```suggestion
		readme.write(`\n### ${userscriptNameFormatter({ baseName, metadata })}\n`);
```

kellnerd reviewed (approved):
> I've performed some very basic testing generating a single userscript but no bookmarklets.

Great job, I've just tested this PR with my repository and the behaviour was as expected ...it deleted my bookmarklets from the README :grin:
Then I've updated my configuration to enable bookmarklets again and strip the leading "MusicBrainz: " from the userscript names. Actually I would have to explicitly title case them to get identical output, but I quite like the way they are capitalised now, so thanks again for bringing this up!

> I'm not sure if the TypeScript definition for UserscriptNameFormatter is needed? Is it possible to define a type in JSDoc comments?

It is possible, but it looks ugly for functions to be honest:
```js
/**
 * Returns the formatted name of the userscript for use in documentation.
 * @typedef {({baseName, metadata}: {baseName: string, metadata: UserscriptMetadata}) => string} UserscriptNameFormatter
 */
```
Also, JSDoc does not allow us to export the type, which might be useful if someone prefers to define a custom name formatter but does not want to have an inline arrow function, just as we did for the default formatter.

> I also changed the default value for bookmarkletSourcePath to null from false, as this seems a more correct way of specifying "no value".

Absolutely, I don't know why I initially chose `false`. Usually it only makes sense to allow that if you can also specify `true`.

Final question: Do you also want to do the related changes to the metadata generator in this PR? Otherwise this is good to merge and I could take care of it.

kellnerd commented on src/types/UserscriptNameFormatter.d.ts:
```diff
@@ -0,0 +1,3 @@
+import { UserscriptMetadata } from "./UserscriptMetadata"
+
+export type UserscriptNameFormatter = ({baseName, metadata}: {baseName: string, metadata: UserscriptMetadata}) => string;
```
```suggestion
export type UserscriptNameFormatter = ({ baseName, metadata }: { baseName: string, metadata: UserscriptMetadata }) => string;
```

atj commented:
OK, I think I've resolved all the comments.

> Final question: Do you also want to do the related changes to the metadata generator in this PR? Otherwise this is good to merge and I could take care of it.

Would you mind taking care of that? I'd prefer to do it myself, but I did take a look before and it wasn't clear to me how best to refactor the interface, but perhaps I was over complicating things.

Thanks for all your help and feedback.

atj committed:
README.md: Update args in example build() invocation

kellnerd commented:
>  I'd prefer to do it myself, but I did take a look before and it wasn't clear to me how best to refactor the interface, but perhaps I was over complicating things.

I have thought to drop the constant `GITHUB` from `github.js` and start as follows in `build.js`:
```js
const gitRepo = GitRepo.fromPackageMetadata({
	// packageJsonPath = 'package.json'
	userscriptNameFormatter,
});

const userscriptNames = await buildUserscripts({
	sourcePath: userscriptSourcePath,
	gitRepo,
	debug,
});
```

And then adapt downstream functions to accept the necessary parameter objects.
The assignment of the homepage URL of course has to happen after the rest of the metadata has already been assigned:
```js
metadata.homepageURL = gitRepo.readmeUrl({ metadata, baseName });
```

atj commented:
OK, I'll have a stab at it then! :monocle_face:

atj committed:
Refactor GitRepo.readmeUrl() to fix README links

Ths change is needed due to the support for customising userscript
documentation section names which was added in a previous commit.

atj commented on src/userscriptMetadata.js:
```diff
@@ -54,6 +55,8 @@ export async function generateMetadataBlock(userscriptPath) {
 		...await loadMetadata(userscriptPath),
 	};
 
+	metadata.homepageURL = gitRepo.readmeSectionUrl({ baseName, metadata });
```
This doesn't feel quite right, but we have a circular dependency - we need an instance of `UserscriptDefaultMetadata` in order to generate the `homepageURL` property.

kellnerd replied:
Yeah, but I don't see a way around this. I think it should be fine as long as nobody expects to be able to construct the section name using the homepage URL. Even in that case we have a sensible default as the empty URL fragment will still be a valid link to the repo.

atj commented:
Well this was a bit more involved than I anticipated!

We now have a mix of functions with and without object destructured  parameters - not sure if you want to standardise and use it everywhere?

Also I renamed `readmeUrl` to `readmeSectionUrl` as it seemed more accurate.

kellnerd committed:
Standardise passing of optional parameters

- Always use object parameters to pass additional options
- Keep required primary parameters outside of these objects
- Create type declarations for parameter objects which are passed on

kellnerd commented:
> We now have a mix of functions with and without object destructured parameters - not sure if you want to standardise and use it everywhere?

Exactly, the idea is to standardise optional parameters to use object destructuring. Having said this, I see that I accidentally suggested to move `sourcePath` into the parameter object although it is required, I've undone that now.

From my side this looks good to merge, I've successfully tested it with my userscript repo and did use a few name formatter variations to see if the metadata generator was working as expected.
